File: src/internal/models/aliases.rs
use serde::Deserialize;
use std::collections::HashMap;
use std::sync::OnceLock;

#[derive(Debug, Clone, Deserialize)]
struct AliasPrefixRule {
    display_prefix: String,
    backend_prefix: String,
}

#[derive(Debug, Clone, Deserialize)]
struct AliasConfig {
    #[serde(default)]
    exact: HashMap<String, String>,
    #[serde(default)]
    prefixes: Vec<AliasPrefixRule>,
}

static ALIAS_CONFIG: OnceLock<AliasConfig> = OnceLock::new();
static REVERSE_EXPLICIT_ALIASES: OnceLock<HashMap<String, String>> = OnceLock::new();

fn alias_config() -> &'static AliasConfig {
    ALIAS_CONFIG.get_or_init(|| {
        let mut cfg: AliasConfig = serde_yaml::from_str(include_str!("model_aliases.yaml"))
            .expect("load model alias config");
        for rule in &mut cfg.prefixes {
            rule.display_prefix = rule.display_prefix.trim().to_string();
            rule.backend_prefix = rule.backend_prefix.trim().to_string();
            if rule.display_prefix.is_empty() || rule.backend_prefix.is_empty() {
                panic!("load model alias config: empty display_prefix/backend_prefix");
            }
        }
        cfg
    })
}

fn reverse_explicit_aliases() -> &'static HashMap<String, String> {
    REVERSE_EXPLICIT_ALIASES.get_or_init(|| {
        alias_config()
            .exact
            .iter()
            .map(|(display, backend)| (backend.clone(), display.clone()))
            .collect()
    })
}

pub fn resolve_model_id(model_id: &str) -> String {
    let model_id = model_id.trim();
    if model_id.is_empty() {
        return String::new();
    }

    if let Some(resolved) = alias_config().exact.get(model_id) {
        return resolved.clone();
    }

    for rule in &alias_config().prefixes {
        if let Some(rest) = model_id.strip_prefix(&rule.display_prefix) {
            return format!("{}{}", rule.backend_prefix, rest);
        }
    }

    model_id.to_string()
}

pub fn display_model_id(model_id: &str) -> String {
    let model_id = model_id.trim();
    if model_id.is_empty() {
        return String::new();
    }

    if let Some(display) = reverse_explicit_aliases().get(model_id) {
        return display.clone();
    }

    for rule in &alias_config().prefixes {
        if let Some(rest) = model_id.strip_prefix(&rule.backend_prefix) {
            return format!("{}{}", rule.display_prefix, rest);
        }
    }

    model_id.to_string()
}

pub fn preferred_model_id(requested_model_id: &str, backend_model_id: &str) -> String {
    let requested_model_id = requested_model_id.trim();
    let backend_model_id = backend_model_id.trim();

    if requested_model_id.is_empty() {
        return display_model_id(backend_model_id);
    }
    if backend_model_id.is_empty() || requested_model_id == backend_model_id {
        return requested_model_id.to_string();
    }
    if resolve_model_id(requested_model_id) == backend_model_id {
        return requested_model_id.to_string();
    }
    display_model_id(backend_model_id)
}

pub fn rewrite_model_id_text(text: &str, backend_model_id: &str, display_model_id: &str) -> String {
    let backend_model_id = backend_model_id.trim();
    let display_model_id = display_model_id.trim();
    if text.is_empty()
        || backend_model_id.is_empty()
        || display_model_id.is_empty()
        || backend_model_id == display_model_id
    {
        return text.to_string();
    }
    text.replace(backend_model_id, display_model_id)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn embedded_alias_config_loaded() {
        assert_eq!(
            alias_config().exact.get("kling_image_o1").unwrap(),
            "kirin_image_omni"
        );
        assert_eq!(alias_config().prefixes.len(), 3);
    }

    #[test]
    fn resolve_and_display_model_id() {
        let cases = [
            ("kling_duration_extension", "kirin_duration_extension"),
            ("kling_image_o1", "kirin_image_omni"),
            ("kling_v2_6_i2v", "kirin_v2_6_i2v"),
            ("seedance_2_0", "spark_dance_v2_0"),
            ("seedream_4_5", "spark_dream_4_5"),
            ("gpt-image-2", "gpt-image-2"),
        ];
        for (display, backend) in cases {
            assert_eq!(resolve_model_id(display), backend);
            assert_eq!(display_model_id(backend), display);
        }
    }

    #[test]
    fn preferred_and_rewrite_model_id() {
        assert_eq!(
            preferred_model_id("seedance_2_0", "spark_dance_v2_0"),
            "seedance_2_0"
        );
        assert_eq!(
            rewrite_model_id_text(
                "model=kirin_v3_t2v endpoint=kirin_v3_t2v",
                "kirin_v3_t2v",
                "kling_v3_t2v"
            ),
            "model=kling_v3_t2v endpoint=kling_v3_t2v"
        );
    }
}

File: internal/models/aliases.go
package models

import (
	_ "embed"
	"fmt"
	"strings"

	"gopkg.in/yaml.v3"
)

type aliasPrefixRule struct {
	DisplayPrefix string `yaml:"display_prefix"`
	BackendPrefix string `yaml:"backend_prefix"`
}

type aliasConfig struct {
	Exact    map[string]string `yaml:"exact"`
	Prefixes []aliasPrefixRule `yaml:"prefixes"`
}

//go:embed model_aliases.yaml
var embeddedAliasConfig []byte

var modelAliasConfig = mustLoadAliasConfig()

var reverseExplicitModelAliases = reverseModelAliases(modelAliasConfig.Exact)

const seaCloudSourcePrefix = "seacloud__"

func ResolveModelID(modelID string) string {
	modelID = strings.TrimSpace(modelID)
	if modelID == "" {
		return ""
	}
	modelID = strings.TrimPrefix(modelID, seaCloudSourcePrefix)

	if resolved, ok := modelAliasConfig.Exact[modelID]; ok {
		return resolved
	}

	for _, rule := range modelAliasConfig.Prefixes {
		if strings.HasPrefix(modelID, rule.DisplayPrefix) {
			return rule.BackendPrefix + strings.TrimPrefix(modelID, rule.DisplayPrefix)
		}
	}

	return modelID
}

func DisplayModelID(modelID string) string {
	modelID = strings.TrimSpace(modelID)
	if modelID == "" {
		return ""
	}

	if display, ok := reverseExplicitModelAliases[modelID]; ok {
		return display
	}

	for _, rule := range modelAliasConfig.Prefixes {
		if strings.HasPrefix(modelID, rule.BackendPrefix) {
			return rule.DisplayPrefix + strings.TrimPrefix(modelID, rule.BackendPrefix)
		}
	}

	return modelID
}

func PreferredModelID(requestedModelID, backendModelID string) string {
	requestedModelID = strings.TrimSpace(requestedModelID)
	backendModelID = strings.TrimSpace(backendModelID)

	if requestedModelID == "" {
		return DisplayModelID(backendModelID)
	}
	if backendModelID == "" || requestedModelID == backendModelID {
		return requestedModelID
	}
	if ResolveModelID(requestedModelID) == backendModelID {
		return requestedModelID
	}
	return DisplayModelID(backendModelID)
}

func RewriteModelIDText(text, backendModelID, displayModelID string) string {
	backendModelID = strings.TrimSpace(backendModelID)
	displayModelID = strings.TrimSpace(displayModelID)

	if text == "" || backendModelID == "" || displayModelID == "" || backendModelID == displayModelID {
		return text
	}
	return strings.ReplaceAll(text, backendModelID, displayModelID)
}

func reverseModelAliases(aliases map[string]string) map[string]string {
	reversed := make(map[string]string, len(aliases))
	for display, backend := range aliases {
		reversed[backend] = display
	}
	return reversed
}

func mustLoadAliasConfig() aliasConfig {
	var cfg aliasConfig
	if err := yaml.Unmarshal(embeddedAliasConfig, &cfg); err != nil {
		panic(fmt.Sprintf("load model alias config: %v", err))
	}
	if cfg.Exact == nil {
		cfg.Exact = map[string]string{}
	}
	for i := range cfg.Prefixes {
		cfg.Prefixes[i].DisplayPrefix = strings.TrimSpace(cfg.Prefixes[i].DisplayPrefix)
		cfg.Prefixes[i].BackendPrefix = strings.TrimSpace(cfg.Prefixes[i].BackendPrefix)
		if cfg.Prefixes[i].DisplayPrefix == "" || cfg.Prefixes[i].BackendPrefix == "" {
			panic("load model alias config: empty display_prefix/backend_prefix")
		}
	}
	return cfg
}
